### PR TITLE
[CIT-571] Add --file and package validation to plugin_apt

### PR DESCRIPTION
**Pull request by makr11st**
## Proposed changes

Split from #444 

This is to add `--file` to existing apt plugin.
It adds also package name and metadata verification.

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: sm/plugins/tedge_apt_plugin/src/module_check.rs [hunk 1]
```diff
@@ -0,0 +1,72 @@
+use crate::error::InternalError;
+use std::path::PathBuf;
+use std::process::{Command, Stdio};
+
+/// check that module_name is in file path
```
Comment (didier-wenzek) [inline sm/plugins/tedge_apt_plugin/src/module_check.rs:5]: The comment is misleading. 

```suggestion
/// check that the path of the package to install is actually shaped as expected by apt
```

File: sm/plugins/tedge_apt_plugin/src/module_check.rs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    }
+}
+
+pub fn validate_package(file_path: &str, contain_args: &[&str]) -> Result<String, InternalError> {
+    let package_metadata = PackageMetadata::try_new(file_path).unwrap();
+
+    if package_metadata.metadata_contains_all(contain_args) {
+        if !filepath_has_extension(file_path) {
+            let file_path = create_link_with_extension(file_path)?;
+            return Ok(file_path);
```
Comment (didier-wenzek) [inline sm/plugins/tedge_apt_plugin/src/module_check.rs:58]: There is something strange here. One checks the file first then we create a symlink to have a file with the extension expected by apt.

Is this ".deb" extension actually required?

File: sm/plugins/tedge_apt_plugin/src/main.rs [hunk 3]
```diff
@@ -73,14 +65,41 @@ fn run(operation: PluginOp) -> Result<ExitStatus, InternalError> {
...
+
+                (None, Some(file_path)) => {
+                    let file_path = validate_package(
+                        file_path,
+                        &[&format!("Package: {}", &module), "Debian package"],
+                    )?;
+                    run_cmd("apt-get", &format!("install --quiet --yes {}", file_path))?
+                }
+
+                _ => {
```
Comment (didier-wenzek) [inline sm/plugins/tedge_apt_plugin/src/main.rs:99]: As a matter of style, I would prefer to have the cases in increasing complexity.

Hence this case first with `(None,None)`, then the case with a version but no file, then a file and no version and last both a  file and a version.

File: sm/plugins/tedge_apt_plugin/src/main.rs [hunk 4]
```diff
@@ -73,14 +65,41 @@ fn run(operation: PluginOp) -> Result<ExitStatus, InternalError> {
...
                     "apt-get",
                     &format!("install --quiet --yes {}={}", module, version),
-                )?
-            } else {
-                run_cmd("apt-get", &format!("install --quiet --yes {}", module))?
+                )?,
+
+                (None, Some(file_path)) => {
+                    let file_path = validate_package(
+                        file_path,
```
Comment (didier-wenzek) [inline sm/plugins/tedge_apt_plugin/src/main.rs:93]: It's not obvious as first read that the file_path is changed with a proper extension added.

**Review by didier-wenzek — COMMENTED** [review]
Good code. My comments are more questions than requests for changes.

The code to check the content of the package is perfect but the way the extension is changed is not clear.

File: sm/plugins/tedge_apt_plugin/src/module_check.rs (hunk 2 above)
Comment (makr11st) [reply sm/plugins/tedge_apt_plugin/src/module_check.rs:58]: Unfortunately it is when using `apt-get`:
```shell
$ sudo apt-get install ./.tmp/rolldice_1.16-1build1_amd64

Reading package lists... Done
E: Unsupported file ./.tmp/rolldice_1.16-1build1_amd64 given on commandline
```

whereas for file with extension:
```shell
$ sudo apt-get install ./.tmp/rolldice_1.16-1build1_amd64.deb 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following NEW packages will be installed:
  rolldice
0 upgraded, 1 newly installed, 0 to remove and 56 not upgraded.
Need to get 0 B/9628 B of archives.
After this operation, 31.7 kB of additional disk space will be used.
Selecting previously unselected package rolldice.
(Reading database ... 73170 files and directories currently installed.)
Preparing to unpack .../rolldice_1.16-1build1_amd64.deb ...
Unpacking rolldice (1.16-1build1) ...
Setting up rolldice (1.16-1build1) ...
Processing triggers for man-db (2.9.1-1) ...
```

File: sm/plugins/tedge_apt_plugin/src/module_check.rs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            .arg("-I")
+            .arg(file_path)
+            .stdout(Stdio::piped())
+            .output()?
+            .stdout)
+    }
+}
+
+pub fn validate_package(file_path: &str, contain_args: &[&str]) -> Result<String, InternalError> {
+    let package_metadata = PackageMetadata::try_new(file_path).unwrap();
```
Comment (PradeepKiruvale) [inline sm/plugins/tedge_apt_plugin/src/module_check.rs:53]: remove unwrap

File: sm/plugins/tedge_apt_plugin/src/main.rs [hunk 6]
```diff
@@ -17,6 +22,8 @@ pub enum PluginOp {
         module: String,
         #[structopt(short = "v", long = "--module-version")]
         version: Option<String>,
+        #[structopt(long = "--file")]
```
[the remark this answers is not in the document]
Comment (makr11st) [reply sm/plugins/tedge_apt_plugin/src/main.rs:25]: Spec didn't define shorthand, I don't mind adding it.

File: tests/PySys/apt_plugin/apt_install_local_file_fail/run.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    when a wrong file_path is provided
+    """
+    _ROLLDICE_URL = "http://ftp.br.debian.org/debian/pool/main/r/rolldice/rolldice_1.16-1+b3_amd64.deb"
+    _path_to_rolldice_binary = None
+    _fake_path_to_rolldice_binary = None
+
+    def setup(self):
+        super().setup()
+        self._download_rolldice_binary(url=self._ROLLDICE_URL)          # downloading the binary
+        self.addCleanupFunction(self.cleanup_remove_rolldice_binary)    # adding cleanup function to remove the binary
```
Comment (didier-wenzek) [inline tests/PySys/apt_plugin/apt_install_local_file_fail/run.py:24]: Why do we download this file that is not used?

**Review by didier-wenzek — APPROVED** [review]
Approved.

Can you please remove the trailing `dbg!` before merging? Thanks.

**Pull request merged**